File: pipeline_ai/ai-engine/app/services/base_service.py
from abc import ABC, abstractmethod
import logging

class BaseAIService(ABC):
    """
    Abstract Base Class cho tất cả các AI Services trong Python Pipeline.
    Đảm bảo tính nhất quán về Logging, Health Status và Lifecycle Management.
    """
    def __init__(self, service_name: str):
        self.service_name = service_name
        self.logger = logging.getLogger(service_name)

    @abstractmethod
    def health_check(self) -> dict:
        """Trả về trạng thái sức khỏe và độ sẵn sàng của Service."""
        pass

    def log_info(self, message: str):
        self.logger.info(f"[{self.service_name}] {message}")

    def log_error(self, message: str, exc: Exception = None):
        if exc:
            self.logger.error(f"[{self.service_name}] {message}: {exc}", exc_info=True)
        else:
            self.logger.error(f"[{self.service_name}] {message}")

File: server/src/common/services/base.service.ts
import { Logger } from '@nestjs/common';

/**
 * Abstract Base Service cung cấp nền tảng Logging & Utility chung cho các NestJS Services.
 */
export abstract class BaseService {
  protected readonly logger: Logger;

  constructor(serviceName: string) {
    this.logger = new Logger(serviceName);
  }

  protected logInfo(message: string): void {
    this.logger.log(message);
  }

  protected logWarn(message: string): void {
    this.logger.warn(message);
  }

  protected logError(message: string, trace?: string): void {
    this.logger.error(message, trace);
  }
}
